File: 1720_decode/decode.rs
/*
 * @Date: 2021-05-06 10:12:45
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-05-06 10:24:07
 */

fn decode(encoded: Vec<i32>, first: i32) -> Vec<i32> {
    let size = encoded.len();
    let mut res: Vec<i32> = vec![0; size + 1];
    res[0] = first;
    for i in 0..size {
        res[i + 1] = encoded[i] ^ res[i];
    }
    res
}

fn main() {
    {
        let encode = vec![1, 2, 3];
        let first = 1;
        assert_eq!(decode(encode, first), vec![1, 0, 2, 1]);
    }
    {
        let encode = vec![6, 2, 7, 3];
        let first = 4;
        assert_eq!(decode(encode, first), vec![4, 2, 0, 7, 4]);
    }
}

File: cpp/1720_decode/decode.cpp
/*
 * @Date: 2021-05-06 10:12:38
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-05-06 10:19:27
 */

#include <cassert>
#include <vector>

using namespace std;

// 若a^b=c,则a=b^c
vector<int> decode(vector<int>& encoded, int first) {
  int m = encoded.size();
  m++;
  vector<int> ans(m);
  ans[0] = first;
  for (auto i = 0; i < encoded.size(); ++i) {
    ans[i + 1] = ans[i] ^ encoded[i];
  }

  return ans;
}

int main() {
  {
    vector<int> encoded{1, 2, 3};
    int first = 1;
    assert(decode(encoded, first) == std::move(vector<int>{1, 0, 2, 1}));
  }
  {
    vector<int> encoded{6, 2, 7, 3};
    int first = 4;
    assert(decode(encoded, first) == std::move(vector<int>{4, 2, 0, 7, 4}));
  }

  return 0;
}